Deps save and drop dependencies

gabrielwol opened the pull request:
update public.deps_save_and_drop_dependencies to include "ALTER OWNER" and "ADD (UNIQUE) CONSTRAINT" statements in deps_saved_ddl table. 
Also fluffed and added to version control!

gabrielwol commented:
@chmnata I think we'll need to use this function for current [miovision re-partitioning ](https://github.com/CityofToronto/bdit_data-sources/pull/735) so your review would be very helpful! 
I tried to organize it so the original function + fluff is in first two commits, so if you just look at the remaining 5 commits it's easier to see just the substantial differences:
https://github.com/CityofToronto/bdit_pgutils/pull/9/files/709cff5ce69a7537cca3efd1abf4235881093f31..2630845fe0415afb19c8dd055be169dcf9b30c3d

chmnata commented:
Looks good! Have you tested the function yet? I think it would be good to add a comment for the function to describe what they do, including the statements it can handle, e.g. alter owners, grant permissions, save table and column comments, index and ......

gabrielwol commented:
@chmnata I added a looong comment. 
I also added a dryrun option that Raph was interested in, which allows a user to run the function without actually dropping deps. 

I tested in my personal schema (see `gwolofs.deps_save_and_drop_dependencies`) and based on that experience also reorganized the inserts so that the CREATE statement comes first, before any modifications (ALTER/GRANT/COMMENT).

chmnata commented:
amazing comment!!! I will test it out with a dry run with a couple of tables, if it looks good we can mergeee :meow_salute:

chmnata commented on create-function-deps_restore_dependencies.sql:
```diff
@@ -0,0 +1,31 @@
+CREATE OR REPLACE FUNCTION public.deps_restore_dependencies(p_view_schema IN VARCHAR, p_view_name IN VARCHAR)
+RETURNS NULL
```
is this supposed to be `void` ? Its not recognizing NULL as an optio. Or is it supposed to be `RETURNS NULL ON NULL INPUT` ?

gabrielwol replied:
right, void. I had to change that on the other one during testing.

chmnata commented on create-function-deps_save_and_drop_dependencies.sql:
```diff
@@ -0,0 +1,212 @@
+CREATE OR REPLACE FUNCTION public.deps_save_and_drop_dependencies(p_view_schema IN VARCHAR, p_view_name IN VARCHAR, dryrun BOOLEAN default False)
```
this one looks good!

chmnata commented on create-function-deps_save_and_drop_dependencies.sql:
```diff
@@ -0,0 +1,212 @@
+END;
+$$;
+
+ALTER FUNCTION public.deps_save_and_drop_dependencies OWNER TO natalie;
```
Can you add (VARCHAR, VARCHAR, BOOLEAN) to the end of the function as well? To make it explicit

gabrielwol commented:
In my testing I found the permissions aren't saved for mat views.
some solutions offered here: https://stackoverflow.com/questions/38728548/list-grants-and-privileges-for-a-materialized-view-in-postgresql

chmnata commented:
Just documenting our IRL discussion and next steps:
- Separate dry run to another function
- Delete existing rows related to the table we are running the function on in the deps table before inserting again

Out of scope in this issue:
- Include create trigger rules

gabrielwol commented:
I changed the main function into deps_save_and_drop_dependencies **_dryrun** with EXECUTE to bdit_humans and made a new function into `deps_save_and_drop_dependencies` with default dryrun = False with permissions only to bdit_admins. I think this is good enough because if a human tried to run _dryrun with dryrun = False they would run into permission issues on the drop statements. 
I also added a delete statement to clear `deps_saved_ddl` at the beginning.

chmnata commented on create-function-deps_save_and_drop_dependencies_dryrun.sql:
```diff
@@ -0,0 +1,228 @@
+END;
+$$;
+
+ALTER FUNCTION public.deps_save_and_drop_dependencies_dryrun(VARCHAR, VARCHAR, BOOLEAN) OWNER TO bdit_admins;
```
Can you change to `dbadmin` instead?

chmnata commented on create-function-deps_save_and_drop_dependencies.sql:
```diff
@@ -0,0 +1,38 @@
+);
+$$;
+
+GRANT EXECUTE ON FUNCTION public.deps_save_and_drop_dependencies(VARCHAR, VARCHAR) TO bdit_admins;
```
Can you change to `dbadmin` instead?

chmnata commented on create-function-deps_restore_dependencies.sql:
```diff
@@ -0,0 +1,31 @@
+END;
+$$;
+
+ALTER FUNCTION public.deps_restore_dependencies(VARCHAR, VARCHAR) OWNER TO natalie;
```
should this still be me ?

gabrielwol replied:
changed to dbadmin

chmnata commented:
I think everything looks good! Do you mind sending me the block of code you use for testing? Or maybe you can show it to me tmr with screenshare.

gabrielwol commented:
FYI Here is the code block I used to test: 

```sql
--DROP MATERIALIZED VIEW gwolofs.employees_test_mat_view;
CREATE MATERIALIZED VIEW gwolofs.employees_test_mat_view AS SELECT * FROM gwolofs.employees;
CREATE UNIQUE INDEX ON gwolofs.employees_test_mat_view USING btree(employee_id);
GRANT ALL ON gwolofs.employees_test_mat_view TO natalie;
COMMENT ON MATERIALIZED VIEW gwolofs.employees_test_mat_view IS 'Test';
COMMENT ON COLUMN gwolofs.employees_test_mat_view.department IS 'Test';

--dry run example: 
SELECT gwolofs.deps_save_and_drop_dependencies_dryrun('gwolofs'::text COLLATE pg_catalog."C", 'employees'::text COLLATE pg_catalog."C");
SELECT deps_ddl_to_run FROM gwolofs.deps_saved_ddl WHERE deps_view_schema = 'gwolofs' AND deps_view_name = 'employees' ORDER BY deps_id;

--not dry run: 
SELECT gwolofs.deps_save_and_drop_dependencies('gwolofs'::text COLLATE pg_catalog."C", 'employees'::text COLLATE pg_catalog."C");
SELECT gwolofs.deps_restore_dependencies('gwolofs'::text COLLATE pg_catalog."C", 'employees'::text COLLATE pg_catalog."C");

/*triggers on views is not supported: 
CREATE OR REPLACE TRIGGER transform_trigger
    INSTEAD OF INSERT
    ON gwolofs.employees_test_mat_view
    FOR EACH ROW
    EXECUTE FUNCTION here.here_insert_trigger();
*/
```

chmnata commented:
Amazing, thanks! Merging
